sklearn/utils/extmath.py

```diff
@@ -1053,12 +1063,33 @@ def _safe_accumulator_op(op, x, *args, **kwargs):
     -------
     result
         The output of the accumulator function passed to this function.
+
+    Notes
+    -----
+    When using array-api support, the accumulator function will upcast floating-point
+    arguments to the maximum precision possible for the array namespace and device.
+    This is usually float64, but may be float32 for some namespace/device pairs.
     """
-    if np.issubdtype(x.dtype, np.floating) and x.dtype.itemsize < 8:
-        result = op(x, *args, **kwargs, dtype=np.float64)
-    else:
-        result = op(x, *args, **kwargs)
-    return result
+    xp, _, x_device = get_namespace_and_device(x)
+    if xp.isdtype(x.dtype, "real floating") and xp.finfo(x.dtype).bits < 64:
```

Comment:
I think it would be more explicit to compare to the max supported dtype size rather than 64 i.e. something like this?
```py
if xp.isdtype(x.dtype, "real floating"):
    max_precision_float_dtype = _max_precision_float_dtype(xp, device=x_device)
    if xp.finfo(xdtype).bits < xp.finfo(max_precision_float_dtype).bits:
        # can upcast
```